### PR TITLE
feat(coordinator): coordinator support multi-circuits prover

### PR DESCRIPTION
**Pull request by amoylan2**
### Purpose or design rationale of this PR

Now coordinator can support prover registering two circuits versions when login, the real version(and hard_fork_name) is decided after get_task is called.
Therefore, for the new prover version, the "ProverVersion" and "HardForkName" in gin context and jwt Claims are no longer valid, all check logic and methods relating to those fields should have a new logic to be compatible

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [ ] No, this PR doesn't involve a new deployment, git tag, docker image tag
- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: common/types/message/multi_auth_msg.go
Comment (georgehao) [inline common/types/message/multi_auth_msg.go]: I think don't need add a new msg type. just remove the auth msg, and rename `legacy_auth_msg` to `auth_msg`

Comment (georgehao) [reply common/types/message/multi_auth_msg.go]: `HardForkName` never needs it. and only pass maximum `proverVersion` to coordinator is ok.

**Comment by colinlyguo** [comment]
We can add a unit test for the following case: 2 provers (one legacy, one multi-libzkp version) can log in get the task, and submit proof. it can ensure backward compatibility before going to the e2e test.

and once this PR is merged, can deploy to sepolia first, check the log for backward compatibility.

File: coordinator/internal/logic/submitproof/proof_receiver.go
Comment (colinlyguo) [inline coordinator/internal/logic/submitproof/proof_receiver.go:139]: No need to change this? using the latest verifier would be sufficient.

File: coordinator/internal/logic/provertask/chunk_prover_task.go
Comment (colinlyguo) [inline coordinator/internal/logic/provertask/chunk_prover_task.go:127]: ditto. looks like `taskCtx.HardForkName` is used in another case?

File: coordinator/internal/logic/provertask/batch_prover_task.go
Comment (colinlyguo) [inline coordinator/internal/logic/provertask/batch_prover_task.go:45]: modify `loadEmbedVK()` as well for next upgrade.

File: coordinator/internal/logic/provertask/prover_task.go
Comment (colinlyguo) [inline coordinator/internal/logic/provertask/prover_task.go]: make it configurable (e.g., multi_version_prover_min_version, then the version can start from v0.0.0 if adding this config) because it can be used in some incompatible upgrades for multi-version provers.

File: coordinator/internal/logic/provertask/prover_task.go
Comment (colinlyguo) [inline coordinator/internal/logic/provertask/prover_task.go]: can also print some info that is related to the prover type here.

File: coordinator/internal/logic/provertask/batch_prover_task.go
Comment (colinlyguo) [inline coordinator/internal/logic/provertask/batch_prover_task.go:92]: `for i := 0; i < 5; i++ {`
can change this back to only one time. but we can leave it to another PR.

File: coordinator/internal/logic/provertask/batch_prover_task.go
Comment (colinlyguo) [inline coordinator/internal/logic/provertask/batch_prover_task.go]:
```suggestion
			log.Error("failed to get version by batch", "error", err.Error())
```

File: coordinator/internal/logic/provertask/batch_prover_task.go
Comment (colinlyguo) [inline coordinator/internal/logic/provertask/batch_prover_task.go:146]: can add a check about `taskCtx.HardForkName` to separate two cases.

**Review by colinlyguo — DISMISSED** [review]

File: coordinator/internal/logic/provertask/batch_prover_task.go
Comment (amoylan2) [reply coordinator/internal/logic/provertask/batch_prover_task.go:92]: this loop time is not concerned with this pr, it's better to modify it in another pr

File: coordinator/internal/logic/provertask/batch_prover_task.go
Comment (amoylan2) [reply coordinator/internal/logic/provertask/batch_prover_task.go:146]: I wonder whether or not it's necessary to check the taskCtx.HardForkName here. Since there are only two code paths calling this function, and there shall be no ambiguity.

File: coordinator/internal/logic/submitproof/proof_receiver.go
Comment (amoylan2) [reply coordinator/internal/logic/submitproof/proof_receiver.go:139]: According to hongfan, the verifier is forward compatible (for batches and ignores chunks temporarily). There may only one verifier here?

File: coordinator/internal/logic/provertask/batch_prover_task.go
Comment (colinlyguo) [reply coordinator/internal/logic/provertask/batch_prover_task.go:146]: make sense.

File: coordinator/internal/logic/provertask/prover_task.go
Comment (georgehao) [reply coordinator/internal/logic/provertask/prover_task.go]: As we discussed, provers should pass the prover version here.

Comment (georgehao) [reply coordinator/internal/logic/provertask/prover_task.go]: `CheckScrollRepoVersion` aims to check whether the prover uses too old version. If you make the hard code, how to check?

File: coordinator/internal/logic/provertask/prover_task.go
Comment (georgehao) [reply coordinator/internal/logic/provertask/prover_task.go]: don't rm the prover version

File: coordinator/internal/logic/provertask/prover_task.go
Comment (georgehao) [inline coordinator/internal/logic/provertask/prover_task.go]: ditto

File: coordinator/internal/logic/submitproof/proof_receiver.go
Comment (georgehao) [reply coordinator/internal/logic/submitproof/proof_receiver.go:139]: need change. the latest verifier can verify multiple version batch, but internal logic also needs the forkname to check alternative Bernoulli/Curie batch.

File: coordinator/internal/logic/provertask/chunk_prover_task.go
Comment (georgehao) [inline coordinator/internal/logic/provertask/chunk_prover_task.go]: why you don't put the `var err error` to the variable group

File: coordinator/internal/logic/provertask/chunk_prover_task.go
Comment (georgehao) [inline coordinator/internal/logic/provertask/chunk_prover_task.go]: delete this line

File: coordinator/internal/logic/provertask/chunk_prover_task.go
Comment (georgehao) [inline coordinator/internal/logic/provertask/chunk_prover_task.go]:
```suggestion
			log.Error("failed to get hard fork name by chunk", "error", err.Error())
```

File: coordinator/internal/logic/provertask/batch_prover_task.go
Comment (georgehao) [inline coordinator/internal/logic/provertask/batch_prover_task.go]:
```suggestion
			log.Error("failed to get hard fork name by batch", "error", err.Error())
```

File: coordinator/internal/logic/provertask/batch_prover_task.go
Comment (amoylan2) [reply coordinator/internal/logic/provertask/batch_prover_task.go:45]: loadEmbedVK has been modified, can you check it then?

File: coordinator/internal/logic/provertask/chunk_prover_task.go
Comment (amoylan2) [reply coordinator/internal/logic/provertask/chunk_prover_task.go]: done

File: coordinator/internal/logic/provertask/chunk_prover_task.go
Comment (amoylan2) [reply coordinator/internal/logic/provertask/chunk_prover_task.go]: done

File: coordinator/internal/logic/provertask/prover_task.go
Comment (amoylan2) [reply coordinator/internal/logic/provertask/prover_task.go]: In line 88, proverVersion has already been checked, Can we remove the second check here?

Comment (georgehao) [reply coordinator/internal/logic/provertask/prover_task.go]: yes, however why here check agin?

**Review by georgehao — DISMISSED** [review]

File: coordinator/internal/logic/provertask/prover_task.go
Comment (amoylan2) [reply coordinator/internal/logic/provertask/prover_task.go]: The origin thought is add another check to ensure the prover is rust-version when dealing with vk arrays. Seems not need, will remove it

**Review by georgehao — DISMISSED** [review]

File: coordinator/internal/logic/submitproof/proof_receiver.go
Comment (colinlyguo) [reply coordinator/internal/logic/submitproof/proof_receiver.go:139]: got it. makes sense.

File: coordinator/internal/types/get_task.go
Comment (colinlyguo) [inline coordinator/internal/types/get_task.go]: we can add some comments here to differentiate prover versions for future maintenance (e.g., deprecate some fields). e.g., used for xxx prover.

**Review by colinlyguo — DISMISSED** [review]
LGTM.

File: coordinator/internal/types/get_task.go
Comment (amoylan2) [reply coordinator/internal/types/get_task.go]: done

**Review by colinlyguo — APPROVED** [review]

**Review by georgehao — APPROVED** [review]

**Pull request merged**
